### PR TITLE
feat: Support decimal groupby

### PR DESCRIPTION
**Pull request by flisky**

### CODE REVIEW

File: crates/polars-core/src/frame/group_by/into_groups.rs [hunk 1]
```diff
@@ -185,6 +185,14 @@ where
                 };
                 num_groups_proxy(ca, multithreaded, sorted)
             },
+            #[cfg(feature = "dtype-decimal")]
+            DataType::Decimal(_, _) => {
+                // Safety: DecimalChunked's first field is ChunkedArray<Int128Type>
```
Comment (ritchie46) [inline crates/polars-core/src/frame/group_by/into_groups.rs:190]: We need this more often. Can you add a `to_primitive` on `DecimalChunked`. I want these pointer casts on a single location.

Comment (flisky) [reply crates/polars-core/src/frame/group_by/into_groups.rs]: `self` is `ChunkedArray<T>`, and we are dispatching based on `dtype()`. `DecimalChunked::to_primitive` doesn't work here, right? 
I've changed the comment as others' `// convince the compiler that we are this type.` for less confusing.

Comment (ritchie46) [reply crates/polars-core/src/frame/group_by/into_groups.rs]: Right, I see.

**Comment by ritchie46** [comment]
Thank you @flisky. Looks good.

**Pull request merged**